File: BEmuJava/bemu/src/main/java/com/bloomberglp/blpapi/ReferenceElementFieldData.java
//------------------------------------------------------------------------------
// <copyright project="BEmu_maven" file="/BEmu_maven/bemu/src/main/java/com/bloomberglp/blpapi/ReferenceElementFieldData.java" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

package com.bloomberglp.blpapi;

import java.util.List;
import java.util.ArrayList;
import java.util.Map;

import com.bloomberglp.blpapi.Name;
import com.bloomberglp.blpapi.Element;
import com.bloomberglp.blpapi.Datetime;
import com.bloomberglp.blpapi.SchemaTypeDefinition;

public class ReferenceElementFieldData extends Element
{
    private final List<Element> _fields;

    ReferenceElementFieldData(Map<String, Object> values)
    {
        this._fields = new ArrayList<Element>();
        for(Map.Entry<String, Object> item : values.entrySet())
        {
        	String itemKey = item.getKey();
        	Object itemValue = item.getValue();
        	
        	if(itemValue instanceof Double)
        	{
        		Element elmDouble = new ReferenceElementDouble(itemKey, (Double)itemValue);
        		this._fields.add(elmDouble);
        	}
        	else if(itemValue instanceof Integer)
        	{
        		Element elmInt = new ReferenceElementInt(itemKey, (Integer)itemValue);
                this._fields.add(elmInt);
        	}
        	else if(itemValue instanceof Datetime)
        	{
        		Element elmDatetime = new ReferenceElementDateTime(itemKey, (Datetime)itemValue);
                this._fields.add(elmDatetime);
        	}
        	else if(itemValue instanceof String)
        	{
        		Element elmString = new ReferenceElementString(itemKey, (String)itemValue);
                this._fields.add(elmString);
        	}
        	else if(itemValue instanceof ReferenceElementArrayChainTickers)
                this._fields.add((ReferenceElementArrayChainTickers)itemValue);
        }
    }
    
    public SchemaTypeDefinition typeDefinition() throws Exception
    {
    	return new SchemaTypeDefinition(this.datatype(), new Name("ReferenceFieldData"));
    }
    
    public Name name() throws Exception
    {
    	return new Name("fieldData");
    }
    
    public int numValues()
    {
    	return 1;
    }
    
    public int numElements()
    {
    	return this._fields.size();
    }
    
    public Element getElement(int index)
    {
    	return this._fields.get(index);
    }
    
    public Element getElement(String name) throws Exception
    {
    	for(int i = 0; i < this._fields.size(); i++)
    	{
    		Element item = this._fields.get(i);
    		if(item.name().toString().equals(name))
    		{
    			return item;
    		}
    	}
    	return super.getElement(name);
    }
    
    protected StringBuilder prettyPrint(int tabIndent) throws Exception
    {
        String tabs = com.bloomberglp.blpapi.IndentType.Indent(tabIndent);
        StringBuilder result = new StringBuilder();

        result.append(String.format("%sfieldData[] = {%s", tabs, System.getProperty("line.separator")));
        
        for(int i = 0; i < this._fields.size(); i++)
        {
        	result.append(this._fields.get(i).prettyPrint(tabIndent + 1));
        }

        result.append(String.format("%s}%s", tabs, System.getProperty("line.separator")));
        return result;
    }
    
    public boolean hasElement(String name) throws Exception
    {
    	return this.hasElement(name, false);
    }
    
    public boolean hasElement(String name, boolean excludeNullElements) throws Exception
    {
    	for(int i = 0; i < this._fields.size(); i++)
    	{
    		if(this._fields.get(i).name().toString().equals(name))
    			return true;
    	}
    	return false;
    }
    
    public double getElementAsFloat64(String name) throws Exception
    {
    	return this.getElement(name).getValueAsFloat64();
    }
    
    public int getElementAsInt32(String name) throws Exception
    {
    	return this.getElement(name).getValueAsInt32();
    }
    
    public long getElementAsInt64(String name) throws Exception
    {
    	return this.getElement(name).getValueAsInt64();
    }
    
    public String getElementAsString(String name) throws Exception
    {
    	return this.getElement(name).getValueAsString();
    }
    
    
    
}

File: BEmu/ReferenceDataRequest/ReferenceElementFieldData.cs
//------------------------------------------------------------------------------
// <copyright project="BEmu_csh" file="ReferenceDataRequest/ReferenceElementFieldData.cs" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

namespace Bloomberglp.Blpapi.ReferenceDataRequest
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;

    internal class ReferenceElementFieldData : Element
    {
        private readonly List<Element> _fields;

        internal ReferenceElementFieldData(Dictionary<string, object> values)
        {
            this._fields = new List<Element>();
            
            foreach (var item in values)
            {
                if (item.Value is double)
                {
                    Element elmDouble = new ReferenceElementDouble(item.Key, (double)item.Value);
                    this._fields.Add(elmDouble);
                }
                else if (item.Value is int)
                {
                    Element elmInt = new ReferenceElementInt(item.Key, (int)item.Value);
                    this._fields.Add(elmInt);
                }
                else if (item.Value is DateTime)
                {
                    Element elmDateTime = new ReferenceElementDateTime(item.Key, (DateTime)item.Value);
                    this._fields.Add(elmDateTime);
                }
                else if (item.Value is string)
                {
                    Element elmString = new ReferenceElementString(item.Key, (string)item.Value);
                    this._fields.Add(elmString);
                }
                else if (item.Value is ReferenceElementArrayChainTickers)
                    this._fields.Add((ReferenceElementArrayChainTickers)item.Value);
            }
        }

        public override IEnumerable<Element> Elements { get { return this._fields; } }

        public override Name Name { get { return new Name("fieldData"); } }
        public override int NumValues { get { return 1; } }
        public override int NumElements { get { return this._fields.Count; } }
        public override SchemaTypeDefinition TypeDefinition { get { return new SchemaTypeDefinition(this.Datatype, new Name("ReferenceFieldData")); } }

        public override Element GetElement(int index)
        {
            return this._fields[index];
        }

        public override Element GetElement(string name)
        {
            return this[name];
        }

        internal override StringBuilder PrettyPrint(int tabIndent, bool surroundValueWithQuotes = false)
        {
            string tabs = Types.IndentType.Indent(tabIndent);
            StringBuilder result = new StringBuilder();

            result.AppendFormat("{0}fieldData = {{{1}", tabs, Environment.NewLine);
            foreach (var item in this._fields)
            {
                result.Append(item.PrettyPrint(tabIndent + 1));
            }
            result.AppendFormat("{0}}}{1}", tabs, Environment.NewLine);
            return result;
        }

        public override bool HasElement(string name, bool excludeNullElements = false)
        {
            foreach (var item in this._fields)
            {
                if (item.Name.ToString() == name)
                    return true;
            }
            return false;
        }

        public override Element this[string name]
        {
            get
            {
                foreach (var item in this._fields)
                {
                    if (item.Name.ToString() == name)
                        return item;
                }
                return base[name];
            }
        }

        public override double GetElementAsFloat64(string name)
        {
            return this[name].GetValueAsFloat64();
        }

        public override int GetElementAsInt32(string name)
        {
            return this[name].GetValueAsInt32();
        }

        public override long GetElementAsInt64(string name)
        {
            return this[name].GetValueAsInt64();
        }

        public override string GetElementAsString(string name)
        {
            return this[name].GetValueAsString();
        }

        public override Datetime GetElementAsDatetime(string name)
        {
            return this[name].GetValueAsDatetime();
        }

    }
}
